R/formula.R

```diff
@@ -0,0 +1,129 @@
+new_googlesheets4_formula <- function(x = character()) {
+  vec_assert(x, character())
+  new_vctr(x, class = "googlesheets4_formula")
+}
+
+#' `googlesheets4_formula` class
+#'
+#' In order to write a formula into Google Sheets, you need to store it as an
+#' object of class `googlesheets4_formula`. This is how we distinguish a
+#' "regular" character string from a string that should be interpreted as a
+#' formula. `googlesheets4_formula` is an S3 class implemented using the [vctrs
+#' package](https://vctrs.r-lib.org/articles/s3-vector.html).
+#'
+#' @param x Character.
+#'
+#' @return An S3 vector of class `googlesheets4_formula`.
+#' @export
+#' @family write functions
+#'
+#' @examples
+#' if (sheets_has_token()) {
+#' dat <- data.frame(X = c(1, 5, 3, 2, 4, 6))
+#'
+#' ss <- sheets_write(dat)
+#'
+#' summaries <- tibble::tribble(
+#'        ~ desc, ~ summaries,
+#'         "max", "=max(A:A)",
+#'         "sum", "=sum(A:A)",
+#'         "min", "=min(A:A)",
+#'   "sparkline", "=SPARKLINE(A:A, {\"color\", \"blue\"})"
+#' )
+#'
+#' # explicitly declare a column as `googlesheets4_formula`
+#' summaries$summaries <- googlesheets4_formula(summaries$summaries)
+#'
+#' sheets_edit(ss, data = summaries, range = "C1", reformat = FALSE)
+#'
+#' miscellany <- tibble::tribble(
+#'        ~ desc, ~ example,
+#'   "hyperlink", "=HYPERLINK(\"http://www.google.com/\",\"Google\")",
+#'        "image", "=IMAGE(\"https://www.google.com/images/srpr/logo3w.png\")"
+#' )
+#' miscellany$example <- googlesheets4_formula(miscellany$example)
+#'
+#' sheets_write(miscellany, ss = ss)
+#'
+#' # clean up
+#' googledrive::drive_trash(ss)
+#' }
+googlesheets4_formula <- function(x = character()) {
```

Comment:
This name still feels too long to me.

Reply:
I agree and ... yet this is walking the talk settled upon here:

https://github.com/tidyverse/googlesheets4/issues/71#issuecomment-564758313

Reply:
I think `sheets_*` was discussed but rejected? This would be consistent with the function naming scheme.

Reply:
class name != constructor name — the class name has to include the package name; the constructor doesn't. If it wasn't already an existing base class, I'd suggest you call the function `formula()`.

Reply:
So user-facing constructor should be: `sheets_formula()`?

We should keep in mind that whatever pattern we use here could also come into play for implementing some of this, i.e. for applying a class to data frame columns that allows us to transmit formats:
<img width="220" alt="Screen Shot 2020-02-22 at 2 29 31 PM" src="https://user-images.githubusercontent.com/599454/75903537-93216a00-5df6-11ea-979e-a4f962e0c10a.png">

Reply:
`sheet_formula()` and `sheets_format()` seem like a nice pairing (although I assume formulas can also have formats?)

Reply:
> although I assume formulas can also have formats

Indeed they can. This will have to be implemented in a way that "layers", for lack of a better word.